File: clients/typescript/agentic-sandbox-client/src/trace-manager.ts
// Copyright 2025 The Kubernetes Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

interface Span {
  isRecording(): boolean;
  setAttribute(key: string, value: string | number | boolean): void;
  end(): void;
}

interface Tracer {
  startSpan(name: string): Span;
  startActiveSpan<T>(name: string, fn: (span: Span) => T): T;
}

interface OtelApi {
  trace: {
    getTracer(name: string): Tracer;
    setSpanInContext(span: Span, ctx: unknown): unknown;
    getSpan(ctx: unknown): Span | undefined;
  };
  context: {
    active(): unknown;
    with<T>(ctx: unknown, fn: () => T): T;
    bind<T extends (...args: unknown[]) => unknown>(ctx: unknown, fn: T): T;
  };
  propagation: {
    inject(ctx: unknown, carrier: Record<string, string>): void;
  };
}

class NoOpSpan implements Span {
  isRecording(): boolean {
    return false;
  }
  setAttribute(_key: string, _value: string | number | boolean): void {}
  end(): void {}
}

class NoOpTracer implements Tracer {
  startSpan(_name: string): Span {
    return new NoOpSpan();
  }
  startActiveSpan<T>(_name: string, fn: (span: Span) => T): T {
    return fn(new NoOpSpan());
  }
}

let otelApi: OtelApi | null = null;
let otelLoaded = false;

async function loadOtel(): Promise<OtelApi | null> {
  if (otelLoaded) return otelApi;
  otelLoaded = true;
  try {
    // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
    const api: Record<string, unknown> = await (
      Function('return import("@opentelemetry/api")')() as Promise<
        Record<string, unknown>
      >
    );
    otelApi = {
      trace: api.trace as OtelApi["trace"],
      context: api.context as OtelApi["context"],
      propagation: api.propagation as OtelApi["propagation"],
    };
    return otelApi;
  } catch {
    return null;
  }
}

let tracerProviderInitialized = false;
let tracerProviderServiceName: string | null = null;

export async function initializeTracer(serviceName: string): Promise<void> {
  const api = await loadOtel();
  if (!api) {
    console.error(
      "OpenTelemetry not installed; skipping tracer initialization.",
    );
    return;
  }

  if (tracerProviderInitialized) {
    if (
      tracerProviderServiceName &&
      tracerProviderServiceName !== serviceName
    ) {
      console.warn(
        `Global TracerProvider already initialized with service name '${tracerProviderServiceName}'. ` +
          `Ignoring request to initialize with '${serviceName}'.`,
      );
    }
    return;
  }

  try {
    const dynamicImport = Function(
      "specifier",
      "return import(specifier)",
    ) as (specifier: string) => Promise<Record<string, unknown>>;

    const sdkTraceNode = await dynamicImport(
      "@opentelemetry/sdk-trace-node",
    );
    const resources = await dynamicImport("@opentelemetry/resources");
    const sdkTraceBase = await dynamicImport(
      "@opentelemetry/sdk-trace-base",
    );
    const exporterOtlpGrpc = await dynamicImport(
      "@opentelemetry/exporter-trace-otlp-grpc",
    );

    const Resource = resources.Resource as new (
      attrs: Record<string, string>,
    ) => unknown;
    const NodeTracerProvider = sdkTraceNode.NodeTracerProvider as new (
      opts: { resource: unknown },
    ) => {
      addSpanProcessor(processor: unknown): void;
      register(): void;
    };
    const BatchSpanProcessor = sdkTraceBase.BatchSpanProcessor as new (
      exporter: unknown,
    ) => unknown;
    const OTLPTraceExporter =
      exporterOtlpGrpc.OTLPTraceExporter as new () => unknown;

    const resource = new Resource({ "service.name": serviceName });
    const provider = new NodeTracerProvider({ resource });
    provider.addSpanProcessor(
      new BatchSpanProcessor(new OTLPTraceExporter()),
    );
    provider.register();

    tracerProviderInitialized = true;
    tracerProviderServiceName = serviceName;
    console.info(
      `Global OpenTelemetry TracerProvider configured for service '${serviceName}'.`,
    );
  } catch {
    console.warn(
      "OpenTelemetry SDK packages not installed; tracer provider not configured. " +
        "Tracing spans will use the default no-op provider.",
    );
  }
}

export async function withSpan<T>(
  tracer: Tracer | null,
  serviceName: string,
  spanSuffix: string,
  fn: (span: Span) => T | Promise<T>,
): Promise<T> {
  if (!tracer) {
    return fn(new NoOpSpan());
  }
  const spanName = `${serviceName}.${spanSuffix}`;
  return tracer.startActiveSpan(spanName, async (span) => {
    try {
      return await fn(span);
    } finally {
      span.end();
    }
  });
}

export function getCurrentSpan(): Span {
  if (!otelApi) return new NoOpSpan();
  const span = otelApi.trace.getSpan(otelApi.context.active());
  return span ?? new NoOpSpan();
}

export class TracerManager {
  public tracer: Tracer;
  private lifecycleSpanName: string;
  private parentSpan: Span | null = null;
  private parentContext: unknown = null;

  constructor(serviceName: string) {
    const scopeName = serviceName.replace(/-/g, "_");
    if (otelApi) {
      this.tracer = otelApi.trace.getTracer(scopeName);
    } else {
      this.tracer = new NoOpTracer();
    }
    this.lifecycleSpanName = `${serviceName}.lifecycle`;
  }

  startLifecycleSpan(): void {
    this.parentSpan = this.tracer.startSpan(this.lifecycleSpanName);
    if (otelApi && this.parentSpan) {
      this.parentContext = otelApi.trace.setSpanInContext(
        this.parentSpan,
        otelApi.context.active(),
      );
    }
  }

  endLifecycleSpan(): void {
    if (this.parentSpan) {
      this.parentSpan.end();
      this.parentSpan = null;
    }
    this.parentContext = null;
  }

  getTraceContextJson(): string {
    if (!otelApi || !this.parentContext) return "";
    const carrier: Record<string, string> = {};
    otelApi.propagation.inject(this.parentContext, carrier);
    return Object.keys(carrier).length > 0 ? JSON.stringify(carrier) : "";
  }
}

export { NoOpSpan, NoOpTracer };
export type { Span, Tracer };

File: clients/python/agentic-sandbox-client/k8s_agent_sandbox/trace_manager.py
# Copyright 2025 The Kubernetes Authors.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""
This module manages OpenTelemetry tracing integration for the Agentic Sandbox client.
It provides a wrapper for the OpenTelemetry SDK to handle tracing initialization,
span creation, and context propagation. If OpenTelemetry is not installed, it
falls back to no-op mock objects.
"""

import atexit
import functools
import json
import logging
from contextlib import nullcontext
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from .models import SandboxTracerConfig

# If optional dependency OpenTelemetry is not installed, define a complete set of mock objects
# to prevent runtime errors.
try:
    from opentelemetry import trace, context
    from opentelemetry.exporter.otlp.proto.grpc.trace_exporter import OTLPSpanExporter
    from opentelemetry.sdk.resources import Resource
    from opentelemetry.sdk.trace import TracerProvider
    from opentelemetry.sdk.trace.export import BatchSpanProcessor
    from opentelemetry.trace.propagation.tracecontext import TraceContextTextMapPropagator
    OPENTELEMETRY_AVAILABLE = True
except ImportError:
    OPENTELEMETRY_AVAILABLE = False
    logging.debug("OpenTelemetry not installed; using MockTracer.")

    class MockSpan:
        """Mock class for OpenTelemetry Span."""

        def is_recording(self):
            """Mock is_recording."""
            return False

        def set_attribute(self, key, value):
            """Mock set_attribute."""

        def end(self):
            """Mock end."""

    class MockTracer:
        """Mock class for OpenTelemetry Tracer."""

        def start_as_current_span(self, *args, **kwargs):
            """Mock start_as_current_span."""
            return nullcontext()

        def start_span(self, *args, **kwargs):
            """Mock start_span."""
            return MockSpan()

    class TraceStub:
        """Mock class for OpenTelemetry trace module."""
        @staticmethod
        def get_current_span():
            """Mock get_current_span."""
            return MockSpan()

        @staticmethod
        def set_tracer_provider(_):
            """Mock set_tracer_provider."""
        @staticmethod
        def get_tracer(name, version=None):
            """Mock get_tracer."""
            return MockTracer()

        @staticmethod
        def set_span_in_context(span, context=None):
            """Mock set_span_in_context."""

    class TraceContextTextMapPropagator:
        """Mock class for OpenTelemetry TraceContextTextMapPropagator."""

        def inject(self, carrier, context=None, setter=None):
            """Mock inject."""

    class ContextStub:
        """Mock class for OpenTelemetry context module."""
        @staticmethod
        def attach(*args, **kwargs):
            """Mock attach."""

        @staticmethod
        def detach(*args, **kwargs):
            """Mock detach."""

    # Assign mock stubs to match import names
    trace = TraceStub
    context = ContextStub



def create_tracer_provider(service_name: str) -> "TracerProvider | None":
    """Creates a TracerProvider with an OTLP/gRPC exporter.

    The endpoint is read from OTEL_EXPORTER_OTLP_ENDPOINT (default: localhost:4317).
    The caller owns the returned provider and should pass it to SandboxTracerConfig.
    """
    if not OPENTELEMETRY_AVAILABLE:
        logging.error(
            "OpenTelemetry not installed; cannot create TracerProvider.")
        return None

    resource = Resource(attributes={"service.name": service_name})
    provider = TracerProvider(resource=resource)
    provider.add_span_processor(
        BatchSpanProcessor(OTLPSpanExporter())
    )
    atexit.register(provider.shutdown)
    return provider


def trace_span(span_suffix):
    """
    Decorator to wrap a method in an OpenTelemetry span with a dynamic name.

    The final span name is constructed at runtime as:
        "{self.trace_service_name}.{span_suffix}"

    The decorated method must belong to an instance (`self`) that has:
        1. `self.tracer`: An initialized OpenTelemetry Tracer.
        2. `self.trace_service_name`: The string name of the service (e.g., 'sandbox-client').

    If `self.tracer` is None (tracing disabled), the method runs without decoration.
    """
    def decorator(func):
        @functools.wraps(func)
        def wrapper(self, *args, **kwargs):
            tracer = getattr(self, 'tracer', None)
            if not tracer:
                return func(self, *args, **kwargs)

            # Determine the service name at runtime
            service_name = getattr(
                self, 'trace_service_name', 'sandbox-client')
            span_name = f"{service_name}.{span_suffix}"

            with tracer.start_as_current_span(span_name):
                return func(self, *args, **kwargs)
        return wrapper
    return decorator


def async_trace_span(span_suffix):
    """
    Async version of trace_span. Wraps an async method in an OpenTelemetry span.

    Same requirements as trace_span: the instance must have `self.tracer` and
    `self.trace_service_name`.
    """
    def decorator(func):
        @functools.wraps(func)
        async def wrapper(self, *args, **kwargs):
            tracer = getattr(self, 'tracer', None)
            if not tracer:
                return await func(self, *args, **kwargs)

            service_name = getattr(
                self, 'trace_service_name', 'sandbox-client')
            span_name = f"{service_name}.{span_suffix}"

            with tracer.start_as_current_span(span_name):
                return await func(self, *args, **kwargs)
        return wrapper
    return decorator


class TracerManager:
    """
    Manages the tracing lifecycle for a single client instance.

    This manager isolates the client's tracing context by:
    1. Creating a tracer with a unique 'instrumentation scope' based on the client's name.
    2. Managing a 'lifecycle' span that serves as the parent for all operations.
    3. Handling the attachment/detachment of the OTel context to the current thread.
    """

    def __init__(self, service_name: str, provider=None):
        instrumentation_scope_name = service_name.replace('-', '_')
        if provider is not None:
            self.tracer = provider.get_tracer(instrumentation_scope_name)
        else:
            self.tracer = trace.get_tracer(instrumentation_scope_name)
        self.lifecycle_span_name = f"{service_name}.lifecycle"
        self.parent_span = None
        self.context_token = None
        self.propagator = TraceContextTextMapPropagator()

    def start_lifecycle_span(self):
        """Starts the main parent span for the client's lifecycle."""
        if not self.tracer:
            return

        self.parent_span = self.tracer.start_span(self.lifecycle_span_name)
        ctx = trace.set_span_in_context(self.parent_span)
        self.context_token = context.attach(ctx)

    def end_lifecycle_span(self):
        """Ends the main parent span and detaches the context."""
        if self.context_token:
            context.detach(self.context_token)
        if self.parent_span:
            self.parent_span.end()

    def get_trace_context_json(self) -> str:
        """Captures only traceparent and tracestate (excludes baggage)."""
        carrier = {}
        self.propagator.inject(carrier)
        return json.dumps(carrier) if carrier else ""

def create_tracer_manager(config: "SandboxTracerConfig", provider=None):
    """Creates a TracerManager from config and an optional TracerProvider."""
    if not config.enable_tracing:
        return None, None

    if not OPENTELEMETRY_AVAILABLE:
        logging.error("OpenTelemetry not installed; skipping tracer initialization.")
        return None, None

    manager = TracerManager(service_name=config.trace_service_name, provider=provider)
    return manager, manager.tracer
